starknet_py/net/contract/contract_function.py

```diff
@@ -0,0 +1,481 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from functools import cached_property
+from typing import List, Optional, Union
+
+from starknet_py.abi import Abi, AbiParser
+from starknet_py.abi.v1.model import Abi as AbiV1
+from starknet_py.abi.v1.parser import AbiParser as AbiV1Parser
+from starknet_py.abi.v2.model import Abi as AbiV2
+from starknet_py.abi.v2.parser import AbiParser as AbiV2Parser
+from starknet_py.abi.v2.shape import L1_HANDLER_ENTRY
+from starknet_py.hash.selector import get_selector_from_name
+from starknet_py.net.account.base_account import BaseAccount
+from starknet_py.net.client import Client
+from starknet_py.net.client_models import Call, EstimatedFee, Hash, ResourceBounds, Tag
+from starknet_py.net.contract.sent_transaction import SentTransaction
+from starknet_py.net.models.transaction import Invoke
+from starknet_py.serialization import (
+    FunctionSerializationAdapter,
+    TupleDataclass,
+    serializer_for_function,
+)
+from starknet_py.serialization.factory import serializer_for_function_v1
+from starknet_py.utils.constructor_args_translator import _is_abi_v2
+from starknet_py.utils.sync import add_sync_methods
+
+ABI = list
+ABIEntry = dict
+
+
+@dataclass(frozen=True)
+class ContractData:
+    """
+    Basic data of a deployed contract.
+    """
+
+    address: int
+    abi: ABI
+    cairo_version: int
+
+    @cached_property
+    def parsed_abi(self) -> Union[Abi, AbiV1, AbiV2]:
+        """
+        Abi parsed into proper dataclass.
+
+        :return: Abi
+        """
+        if self.cairo_version == 1:
+            if _is_abi_v2(self.abi):
+                return AbiV2Parser(self.abi).parse()
+            return AbiV1Parser(self.abi).parse()
+        return AbiParser(self.abi).parse()
+
+    @staticmethod
+    def from_abi(address: int, abi: ABI, cairo_version: int = 0) -> ContractData:
+        """
+        Create ContractData from ABI.
+
+        :param address: Address of the deployed contract.
+        :param abi: Abi of the contract.
+        :param cairo_version: Version of the Cairo in which contract is written.
+        :return: ContractData instance.
+        """
+        return ContractData(
+            address=address,
+            abi=abi,
+            cairo_version=cairo_version,
+        )
+
+
+@add_sync_methods
+@dataclass(frozen=True)
+class InvokeResult(SentTransaction):
+    """
+    Result of the Invoke transaction.
+    """
+
+    # We ensure these are not None in __post_init__
+    contract: ContractData = None  # pyright: ignore
+    """Additional information about the Contract that made the transaction."""
+
+    invoke_transaction: Invoke = None  # pyright: ignore
+    """A InvokeTransaction instance used."""
+
+    def __post_init__(self):
+        assert self.contract is not None
+        assert self.invoke_transaction is not None
+
+
+@dataclass
+class CallToSend(Call):
```

Comment:
Have you consider any other names for this one? 😅 

I'm thinking `PreparedCall` or `PreparedCallBase`.

Reply:
Will be changed to `PrepareCallBase` 👍